fix: developer getting started docs with working example

devigned opened the pull request:
**What this PR does / why we need it**:

The developer getting started docs are really out of date. When trying to follow `./docs/devepment.md`, I ran across several broken commands and missing guidance.

In this PR, I address getting an initial example cluster up and running.
- add kustomize to the required tooling
- remove dependency on `clusterctl` for `make cluster-create`
- add information about generating and cleaning up example manifests
- add wait until ready script to provide status updates while machines are provisioning
- remove content about go vendor directories
- output a `.env` file during example generation to help with commands that may need to be run later

TODOs:
- `make test` doesn't work on my local machine. Seems like `./scripts/fetch_ext_bins.sh` is only called during CI runs and is not really described for local development. Perhaps, make test should execute tests in a container and replicate the CI infra? See below for error.

```
Running Suite: Controller Suite
===============================
Random Seed: 1574273964
Will run 2 of 2 specs

E1120 10:19:24.496084   37823 server.go:228] controller-runtime/test-env "msg"="unable to start the controlplane" "error"="fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory"  "tries"=0
E1120 10:19:24.498637   37823 server.go:228] controller-runtime/test-env "msg"="unable to start the controlplane" "error"="fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory"  "tries"=1
E1120 10:19:24.501045   37823 server.go:228] controller-runtime/test-env "msg"="unable to start the controlplane" "error"="fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory"  "tries"=2
E1120 10:19:24.503288   37823 server.go:228] controller-runtime/test-env "msg"="unable to start the controlplane" "error"="fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory"  "tries"=3
E1120 10:19:24.505579   37823 server.go:228] controller-runtime/test-env "msg"="unable to start the controlplane" "error"="fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory"  "tries"=4
STEP: bootstrapping test environment
Failure [0.015 seconds]
[BeforeSuite] BeforeSuite
/Users/davidjustice/code/k8s/capz/controllers/suite_test.go:59

  Unexpected error:
      <*errors.errorString | 0xc000496bb0>: {
          s: "failed to start the controlplane. retried 5 times: fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory",
      }
      failed to start the controlplane. retried 5 times: fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory
  occurred

  /Users/davidjustice/code/k8s/capz/controllers/suite_test.go:69
------------------------------
STEP: tearing down the test environment
Panic [0.000 seconds]
[AfterSuite] AfterSuite
/Users/davidjustice/code/k8s/capz/controllers/suite_test.go:84

  Test Panicked
  runtime error: invalid memory address or nil pointer dereference
``` 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Update developer getting started documentation.
```

devigned commented:
I signed it

devigned commented:
/check-cla

devigned commented:
/check-cla

CecileRobertMichon commented on docs/development.md:
````diff
@@ -124,13 +117,17 @@ using:
 
 ```bash
 make clean # Clean up any previous manifests
+make clean-examples # Clean up any previous example manifests
````
this isn't strictly needed since `make generate-examples` runs it already.

devigned replied:
👍

CecileRobertMichon commented on docs/development.md:
```diff
@@ -171,9 +155,9 @@ time kubectl get po -o wide --all-namespaces -w # Watch pod creation until azure
-a kubeconfig copied locally. You can debug most issues by SSHing into the
-instances that have been created and reading `/var/lib/waagent/custom-script/download/0/stdout`.
+After this is finished you will have kubeconfig in `./examples/_out/clusterapi.kubeconfig`.
+You can debug most issues by SSHing into the instances that have been created and 
```
s/instances/VMs

devigned replied:
👍

CecileRobertMichon commented on examples/wait-for-ready.sh:
```diff
@@ -0,0 +1,37 @@
+  kubectl --kubeconfig="$kubeconfigPath" get machines
+  output=$(kubectl --kubeconfig="$kubeconfigPath" get machines -o json)
+  ready=$(echo "$output" | jq -r 'all(.items[]; .status.phase=="running")')
+  [ "$ready" == "false" ] && sleep 60
```
I think we can probably make this retry loop a bit shorter as we discussed

CecileRobertMichon replied:
(without printing the logs every time)

devigned replied:
👍

CecileRobertMichon commented on Makefile:
```diff
@@ -305,31 +294,38 @@ create-cluster-management: $(CLUSTERCTL) ## Create a development Kubernetes clus
-		--kubeconfig=./kubeconfig \
-		-a examples/addons.yaml
 	# Create a worker node with MachineDeployment.
 	kubectl \
```
would it make sense to move the worker node create to _after_ the CNI is applied since we don't need the machines to be created to fetch the kubeconfig?

CecileRobertMichon replied:
or maybe change the wait script to not wait for all machines

CecileRobertMichon replied:
that way we do start the create asap

CecileRobertMichon commented on Makefile:
```diff
@@ -305,31 +294,38 @@ create-cluster-management: $(CLUSTERCTL) ## Create a development Kubernetes clus
-		alpha phases apply-addons -v=10 \
-		--kubeconfig=./kubeconfig \
-		-a examples/addons.yaml
 	# Create a worker node with MachineDeployment.
```
same comment for `# Create control plane machine` --> `# Create control plane machines.`

devigned replied:
fixed

CecileRobertMichon commented on Makefile:
```diff
@@ -305,31 +294,38 @@ create-cluster-management: $(CLUSTERCTL) ## Create a development Kubernetes clus
+      apply -f https://docs.projectcalico.org/v3.8/manifests/calico.yaml
+
+	@echo 'run "kubectl --kubeconfig=$$(kind get kubeconfig-path --name="clusterapi") ..." to work with the kind cluster'
+	@echo 'run "kubectl --kubeconfig=./examples/_out/clusterapi.kubeconfig ..." to work with the new cluster'
```
```suggestion
	@echo 'run "kubectl --kubeconfig=./examples/_out/clusterapi.kubeconfig ..." to work with the new target cluster'
```

devigned replied:
fixed

CecileRobertMichon commented on Makefile:
```diff
@@ -305,31 +294,38 @@ create-cluster-management: $(CLUSTERCTL) ## Create a development Kubernetes clus
+	kubectl --kubeconfig=./examples/_out/clusterapi.kubeconfig \
+      apply -f https://docs.projectcalico.org/v3.8/manifests/calico.yaml
+
+	@echo 'run "kubectl --kubeconfig=$$(kind get kubeconfig-path --name="clusterapi") ..." to work with the kind cluster'
```
should this be 
```suggestion
	@echo 'run "kubectl --kubeconfig=$(kind get kubeconfig-path --name="clusterapi")" to work with the kind cluster'
```

devigned replied:
This is how I see the above on the cmd line output:
```
run "kubectl --kubeconfig=$(kind get kubeconfig-path --name="clusterapi") ..." to work with the kind cluster
```

The extra `$` is eaten up by `make`.

CecileRobertMichon replied:
yeah you're right, what about the `...` at the end?

devigned commented:
@CecileRobertMichon I think this last round of changes has addressed your feedback. Let me know if anything else jumps out at you.

```
$ make create-cluster
kind create cluster --name=clusterapi
Creating cluster "clusterapi" ...
 ✓ Ensuring node image (kindest/node:v1.15.3) 🖼
 ✓ Preparing nodes 📦
 ✓ Creating kubeadm config 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
Cluster creation complete. You can now use the cluster with:

export KUBECONFIG="$(kind get kubeconfig-path --name="clusterapi")"
kubectl cluster-info
# Apply provider-components.
kubectl \
		--kubeconfig=$(kind get kubeconfig-path --name="clusterapi") \
		create -f examples/_out/provider-components.yaml
namespace/cabpk-system created
namespace/capi-system created
namespace/capz-system created
customresourcedefinition.apiextensions.k8s.io/azureclusters.infrastructure.cluster.x-k8s.io created
customresourcedefinition.apiextensions.k8s.io/azuremachines.infrastructure.cluster.x-k8s.io created
customresourcedefinition.apiextensions.k8s.io/azuremachinetemplates.infrastructure.cluster.x-k8s.io created
customresourcedefinition.apiextensions.k8s.io/clusters.cluster.x-k8s.io created
customresourcedefinition.apiextensions.k8s.io/kubeadmconfigs.bootstrap.cluster.x-k8s.io created
customresourcedefinition.apiextensions.k8s.io/kubeadmconfigtemplates.bootstrap.cluster.x-k8s.io created
customresourcedefinition.apiextensions.k8s.io/machinedeployments.cluster.x-k8s.io created
customresourcedefinition.apiextensions.k8s.io/machines.cluster.x-k8s.io created
customresourcedefinition.apiextensions.k8s.io/machinesets.cluster.x-k8s.io created
role.rbac.authorization.k8s.io/cabpk-leader-election-role created
role.rbac.authorization.k8s.io/capi-leader-election-role created
role.rbac.authorization.k8s.io/capz-leader-election-role created
clusterrole.rbac.authorization.k8s.io/cabpk-manager-role created
clusterrole.rbac.authorization.k8s.io/cabpk-proxy-role created
clusterrole.rbac.authorization.k8s.io/capi-manager-role created
clusterrole.rbac.authorization.k8s.io/capz-manager-role created
clusterrole.rbac.authorization.k8s.io/capz-proxy-role created
rolebinding.rbac.authorization.k8s.io/cabpk-leader-election-rolebinding created
rolebinding.rbac.authorization.k8s.io/capi-leader-election-rolebinding created
rolebinding.rbac.authorization.k8s.io/capz-leader-election-rolebinding created
clusterrolebinding.rbac.authorization.k8s.io/cabpk-manager-rolebinding created
clusterrolebinding.rbac.authorization.k8s.io/cabpk-proxy-rolebinding created
clusterrolebinding.rbac.authorization.k8s.io/capi-manager-rolebinding created
clusterrolebinding.rbac.authorization.k8s.io/capz-manager-rolebinding created
clusterrolebinding.rbac.authorization.k8s.io/capz-proxy-rolebinding created
secret/capz-manager-bootstrap-credentials created
service/cabpk-controller-manager-metrics-service created
service/capz-controller-manager-metrics-service created
deployment.apps/cabpk-controller-manager created
deployment.apps/capi-controller-manager created
deployment.apps/capz-controller-manager created
# Create Cluster.
kubectl \
		--kubeconfig=$(kind get kubeconfig-path --name="clusterapi") \
		create -f examples/_out/cluster.yaml
cluster.cluster.x-k8s.io/capz-example-a73d1a6d created
azurecluster.infrastructure.cluster.x-k8s.io/capz-example-a73d1a6d created
# Create control plane machines.
kubectl \
		--kubeconfig=$(kind get kubeconfig-path --name="clusterapi") \
		create -f examples/_out/controlplane.yaml
kubeadmconfig.bootstrap.cluster.x-k8s.io/capz-example-a73d1a6d-controlplane-0 created
kubeadmconfig.bootstrap.cluster.x-k8s.io/capz-example-a73d1a6d-controlplane-1 created
kubeadmconfig.bootstrap.cluster.x-k8s.io/capz-example-a73d1a6d-controlplane-2 created
machine.cluster.x-k8s.io/capz-example-a73d1a6d-controlplane-0 created
machine.cluster.x-k8s.io/capz-example-a73d1a6d-controlplane-1 created
machine.cluster.x-k8s.io/capz-example-a73d1a6d-controlplane-2 created
azuremachine.infrastructure.cluster.x-k8s.io/capz-example-a73d1a6d-controlplane-0 created
azuremachine.infrastructure.cluster.x-k8s.io/capz-example-a73d1a6d-controlplane-1 created
azuremachine.infrastructure.cluster.x-k8s.io/capz-example-a73d1a6d-controlplane-2 created
# wait for the first control plane machine to be ready
./examples/wait-for-ready.sh
Waiting for any machines to be in running state in cluster capz-example-a73d1a6d...
Waiting...
Waiting...
Waiting...
Waiting...
Waiting...
Waiting...
Waiting...
Waiting...
Waiting...
Waiting...
Machines are ready
# Fetch the Kubeconfig for the target cluster
source ./examples/_out/.env; \
	kubectl \
		--kubeconfig=$(kind get kubeconfig-path --name="clusterapi") \
		--namespace=default get secret/$CLUSTER_NAME-kubeconfig -o json \
		| jq -r .data.value \
		| base64 --decode > ./examples/_out/clusterapi.kubeconfig
# Deploy a CNI soution, Calico
kubectl --kubeconfig=./examples/_out/clusterapi.kubeconfig \
	  apply -f https://docs.projectcalico.org/v3.8/manifests/calico.yaml
configmap/calico-config created
  1 pick 16a05667 fix: correct developer getting started docs¬
  5 # This is a combination of 2 commits.¬
customresourcedefinition.apiextensions.k8s.io/felixconfigurations.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/ipamblocks.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/blockaffinities.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/ipamhandles.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/ipamconfigs.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/bgppeers.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/bgpconfigurations.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/ippools.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/hostendpoints.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/clusterinformations.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/globalnetworkpolicies.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/globalnetworksets.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/networkpolicies.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/networksets.crd.projectcalico.org created
clusterrole.rbac.authorization.k8s.io/calico-kube-controllers created
clusterrolebinding.rbac.authorization.k8s.io/calico-kube-controllers created
clusterrole.rbac.authorization.k8s.io/calico-node created
clusterrolebinding.rbac.authorization.k8s.io/calico-node created
daemonset.apps/calico-node created
serviceaccount/calico-node created
deployment.apps/calico-kube-controllers created
serviceaccount/calico-kube-controllers created
# Create 2 worker nodes with MachineDeployment.
kubectl \
		--kubeconfig=$(kind get kubeconfig-path --name="clusterapi") \
		create -f examples/_out/machinedeployment.yaml
kubeadmconfigtemplate.bootstrap.cluster.x-k8s.io/capz-example-a73d1a6d-md-0 created
machinedeployment.cluster.x-k8s.io/capz-example-a73d1a6d-md-0 created
azuremachinetemplate.infrastructure.cluster.x-k8s.io/capz-example-a73d1a6d-md-0 created
run "kubectl --kubeconfig=$(kind get kubeconfig-path --name="clusterapi") ..." to work with the kind cluster
run "kubectl --kubeconfig=./examples/_out/clusterapi.kubeconfig ..." to work with the new target cluster
```

CecileRobertMichon commented on examples/wait-for-ready.sh:
```diff
@@ -0,0 +1,33 @@
+  ready=$(echo "$output" | jq -r 'any(.items[]; .status.phase=="running")')
+  [ "$ready" == "false" ] && echo "Waiting..." && sleep 30
+done
+echo 'Machines are ready'
```
```suggestion
echo 'found a ready Machine'
```

devigned replied:
Changed to: `echo 'found a running Machine'`

CecileRobertMichon commented on examples/wait-for-ready.sh:
```diff
@@ -0,0 +1,33 @@
+
+ready="false"
+kubeconfigPath=$(kind get kubeconfig-path --name="clusterapi")
+echo "Waiting for any machines to be in running state in cluster ${CLUSTER_NAME}..."
```
```suggestion
echo "Waiting for any machine to be in running state in cluster ${CLUSTER_NAME}..."
```

devigned replied:
Changed to: `Waiting for any machine in cluster ${CLUSTER_NAME} to be in running state...`

CecileRobertMichon commented:
/lgtm
/approve

thanks @devigned!
